tls: Add CRL support

adunham-stripe opened the pull request:
*Description*:
Adds support for CRLs.  Fixes issue #1496.  Duplicate version of #2077.

*API Changes*:
See Data Plane PR: https://github.com/envoyproxy/data-plane-api/pull/358

*Risk Level*: Medium (TLS is scary!)

*Testing*:
Wrote some automated tests that test that revoked certs can't connect while non-revoked certs can.

PiotrSikora commented on include/envoy/ssl/context_config.h:
```diff
@@ -53,6 +55,11 @@ class ContextConfig {
+  /**
+   * @return The CRL to check if a cert is revoked.
+   */
+  virtual const std::vector<bssl::UniquePtr<X509_CRL>>& certificateRevocationLists() const PURE;
```
This is (common?) interface and not implementation, so I don't think you should use BoringSSL-specific types here.

adunham-stripe replied:
So, there was a request [in the previous PR](https://github.com/envoyproxy/envoy/pull/2077#discussion_r151738924) to have the class hold a `bssl::UniquePtr<X509_CRL>` so that invalid data would be caught sooner.  Agreed that it's a bit ugly to have BoringSSL-specific types in the interface, so happy to make whichever changes you / @ggreenway would prefer.

mattklein123 replied:
I agree with @PiotrSikora that we shouldn't expose boringssl at this interface level. I haven't fully reviewed the change but can we find a way to hide this? We might need to add a CRL interface also.

adunham-stripe replied:
So, here's three possible options:

1. Have `ContextConfig` contain a `DataSource`, or just the raw bytes for a CRL.
  **Pros:** we don't expose BoringSSL in this interface; simple.
  **Cons:** if the CRL is invalid, we won't find out until we try to instantiate a listener (since that's when we turn the bytes into a CRL).

2. Have `ContextConfig` contain the BoringSSL types.
  **Pros:** we parse the CRL ~immediately and thus can error at config load time; already exists.
  **Cons:** doesn't feel like `ContextConfig` should have BoringSSL types.

3. Introduce a `CRL` interface (or something similar), parse the CRL up-front, wrap into the interface and use that here.
  **Pros:** probably can get the best of both worlds (verifying the CRL up-front and hiding the concrete type).
  **Cons:** complicated; need to make sure that whatever parses the config does so into a format that the `ContextConfig` implementation understands (i.e. there's coupling between config loading and the `ContextConfigImpl`).

Thoughts?

mattklein123 replied:
Yeah agreed this is a little tricky. Some random ideas:

1) Variant of (1) where we expose the raw bytes, but also verify that it's valid inside the config and throw if not. (Basically we would end up loading the CRL twice, but I think that's probably not a big deal).

2) Some type of visitor pattern, where the `ContextImpl` calls back into the config and the config sets up the SSL context. This would require some dynamic_cast to get at the `SSL` inside the config impl, but we are already doing that in various places inside config impl for similar reasons, so I think it's probably OK.

PiotrSikora replied:
FWIW, https://github.com/envoyproxy/envoy/pull/2248 is just loading and exposing raw bytes, without verifying the content. We could start loading them twice, but this will start being costly at scale, and will hurt loading times.

ggreenway replied:
I think the most pragmatic solution is to just expose raw bytes in this interface, and catch any data-validation errors in the ContextImpl.  We already catch many similar types of errors there.

PiotrSikora commented on source/common/json/config_schemas.cc:
```diff
@@ -156,7 +156,8 @@ const std::string Json::Schema::LISTENER_SCHEMA(R"EOF(
           "cipher_suites" : {"type" : "string", "minLength" : 1},
-          "ecdh_curves" : {"type" : "string", "minLength" : 1}
+          "ecdh_curves" : {"type" : "string", "minLength" : 1},
+          "crl_file" : {"type" : "string"}
```
v1 APIs are deprecated, so I'm not sure if we're supposed to keep adding things here? cc @mattklein123 @htuch

mattklein123 replied:
At this point v2 is required, v1 is optional but not blocked. If folks want to add and doc no problem.

PiotrSikora commented on source/common/ssl/context_config_impl.cc:
```diff
@@ -34,6 +34,61 @@ const std::string ContextConfigImpl::DEFAULT_CIPHER_SUITES =
+  bssl::UniquePtr<STACK_OF(X509_INFO)> xis(nullptr);
+
+  auto crl_conf = vctx.crl();
+  if (crl_conf.filename().length()) {
```
You should probably wait for `ContextConfigImpl::readDataSource()` from https://github.com/envoyproxy/envoy/pull/2248.

ggreenway replied:
+1

If not though, change to "if (!crl_conf.filename().empty()) {"

PiotrSikora commented on source/common/ssl/context_config_impl.cc:
```diff
@@ -34,6 +34,61 @@ const std::string ContextConfigImpl::DEFAULT_CIPHER_SUITES =
+
+  // Iterate through every item in the stack, adding all found CRLs to our
+  // return array.
+  for (unsigned int i = 0; i < sk_X509_INFO_num(xis.get()); i++) {
```
You might want to look at similar code in https://github.com/envoyproxy/envoy/pull/2248, and use the list iterator from `bssl::UniquePtr`.

adunham-stripe replied:
Ooh, that's good to know - I didn't realize there was an iterator on `STACK`s.  Also, regarding the similar code: as I read it, #2248 will load CRLs from the file passed as a CA cert; I don't want to duplicate work, but $WORK definitely needs to be able to load CRLs from a different file from CAs (we essentially fetch CRLs on a cronjob, whereas the CA certs are placed with config management).  What are your thoughts on reducing duplication here?

PiotrSikora replied:
Yeah, CRLs will be loaded from `trusted_ca` as well, since that's what the filesystem-based API is currently doing, and I didn't want to change the behavior and suddenly start ignoring CRLs from the `trusted_ca` file... Having said that, I don't think that's a common use case and handling CRLs via separate field is definitely OK.

PiotrSikora commented on source/common/ssl/context_config_impl.cc:
```diff
@@ -34,6 +34,61 @@ const std::string ContextConfigImpl::DEFAULT_CIPHER_SUITES =
+      // We need to set the pointer to nullptr here, since otherwise the
+      // pointer's destructor (`sk_X509_INFO_pop_free`) will free the
+      // underlying CRL, that we've just copied to the vector.
+      xi->crl = nullptr;
```
You should use `X509_CRL_up_ref()` instead of this hack.

adunham-stripe replied:
Thanks for the tip - switched to that!

mattklein123 commented:
@ggreenway do you mind also taking a look?

ggreenway commented on bazel/repository_locations.bzl:
```diff
@@ -67,8 +67,8 @@ REPOSITORY_LOCATIONS = dict(
-        commit = "3deffbd132b40fa544137902fdf83bec7442a87f",
-        remote = "https://github.com/envoyproxy/data-plane-api",
+        commit = "ff059f8ae595a262d85dceee09fab1f27738c944",
+        remote = "https://github.com/adunham-stripe/data-plane-api",
```
Reminder: Point at your real data-plane-api commit

adunham-stripe replied:
Done, thanks!  I just pointed at the current commit on `master`.

adunham-stripe replied:
(rebased to fix merge conflict)

ggreenway commented on test/common/ssl/context_impl_test.cc:
```diff
@@ -266,5 +266,17 @@ TEST_F(SslServerContextImplTicketTest, TicketKeyInlineFailTooSmall) {
   EXPECT_THROW(loadConfigV2(cfg), EnvoyException);
 }
 
+TEST_F(SslServerContextImplTicketTest, CRLSuccess) {
```
Please also add a test case for invalid crl (empty file maybe?)

adunham-stripe replied:
Done!  I think an empty file is actually okay (list with 0 elements), so I added a file with invalid PEM data.

adunham-stripe commented:
Okay, I believe that I've made all the review changes that have been requested, except for using `ContextConfigImpl::readDataSource()` and putting the CRL bytes in the context.  I'll wait for #2248 to land, and then rebase on that and use `readDataSource` to read the CRL into the context as bytes.

ggreenway commented:
#2248 is waiting on some other things, so it will be at least another week.  @adunham-stripe would you rather wait on that, or proceed without it?  Or maybe you can check with @PiotrSikora and just add readDataSource() in this PR to unblock this?

adunham-stripe commented:
I'm happy to take either approach!  We don't have an immediate need for CRLs yet, so I'm okay waiting and seeing how #2248 lands first.  Though, if you'd rather get this PR moving sooner, and if @PiotrSikora is onboard, happy to add `readDataSource()` here too.

ggreenway commented:
I'm fine with however you want to proceed; I don't think anyone is blocked on CRLs.  I just wanted to let you know the status of things and give you options.

adunham-stripe commented:
Thank you! 😁   I'm going to hold off for now, then; I've subscribed to the other PR and will follow up after it gets merged 👍

mattklein123 commented:
@adunham-stripe the other PR merged if you want to finish this up. Thank you!

adunham-stripe commented:
Yep, saw that - nice job on the quick turnaround from you folks!  I've rebased on master and am just running tests on the changes that @ggreenway requested.  Hoping to push the changes by EOD.

adunham-stripe commented:
@mattklein123 - Okay, I just pushed the changes that @ggreenway requested in their review, and fixed tests.  Let me know if there's anything else you folks would like!

mattklein123 commented on source/common/config/tls_context_json.cc:
```diff
@@ -61,6 +61,11 @@ void TlsContextJson::translateCommonTlsContext(
     common_tls_context.mutable_tls_params()->add_cipher_suites(std::string{cipher_suite});
   }
 
+  // TODO: this doesn't currently support inline data
```
We won't ever support inline in v1. I would remove this TODO. Did you add docs for v1 also in data-plane-api? I can't remember.

adunham-stripe replied:
Removed!  And yes, I did add documentation [here][docs].

[docs]: https://github.com/envoyproxy/data-plane-api/blob/master/docs/root/api-v1/listeners/listeners.rst#user-content-config-listener-ssl-context-crl-file

PiotrSikora commented on source/common/ssl/context_impl.cc:
```diff
@@ -152,6 +152,30 @@ ContextImpl::ContextImpl(ContextManagerImpl& parent, Stats::Scope& scope,
+          fmt::format("Failed to load CRL from {}", config.certificateRevocationListPath()));
+    }
+
+    X509_STORE* cs = SSL_CTX_get_cert_store(ctx_.get());
```
Nit: `s/cs/store_ctx/`

PiotrSikora commented on source/common/ssl/context_impl.cc:
```diff
@@ -152,6 +152,30 @@ ContextImpl::ContextImpl(ContextManagerImpl& parent, Stats::Scope& scope,
+      }
+    }
+
+    X509_STORE_set_flags(cs, X509_V_FLAG_CRL_CHECK);
```
`X509_V_FLAG_CRL_CHECK|X509_V_FLAG_CRL_CHECK_ALL`

adunham-stripe replied:
Do we actually want to enable `X509_V_FLAG_CRL_CHECK_ALL`?  I had strange failures when turning that flag on during testing, and there's [at least one bug][bug] still open about it.

[bug]: https://github.com/openssl/openssl/issues/5081

PiotrSikora replied:
I'm pretty sure we want it. What were the error, do you recall?

PiotrSikora commented on test/common/ssl/context_impl_test.cc:
```diff
@@ -284,5 +284,29 @@ TEST_F(SslServerContextImplTicketTest, TicketKeyInlineStringFailTooSmall) {
+  }
+  )EOF";
+
+  EXPECT_THROW(loadConfigJson(json), EnvoyException);
```
Could you replace this with `EXPECT_THROW_WITH_MESSAGE` and proper message, to make sure it fails because of invalid CA and not for other reason (like CRL defined, but no trusted CA)?

PiotrSikora commented on test/common/ssl/context_impl_test.cc:
```diff
@@ -284,5 +284,29 @@ TEST_F(SslServerContextImplTicketTest, TicketKeyInlineStringFailTooSmall) {
+  )EOF";
+
+  EXPECT_THROW(loadConfigJson(json), EnvoyException);
+}
```
Could you add test for v2 with inlined CRL as well? See similar tests in `test/server/listener_manager_impl_test.cc `, (...which probably should be moved here instead).

PiotrSikora commented on test/common/ssl/test_data/crl.cfg:
```diff
@@ -0,0 +1,18 @@
+[ default ]
```
Could we merge this into `ca_cert.cfg`?

adunham-stripe replied:
I am honestly not an expert when it comes to OpenSSL config files - I'm happy to take a stab at it if you'd like, but it'll take me a bit to familiarize myself with things, I suspect.

PiotrSikora replied:
I think moving content of this file into `ca_cert.cfg` should work, but skip `dir` and adjust `database` and `crlnumber` to avoid messing up other sub-commands.

If it doesn't "just work", then feel free to leave it as-is.

adunham-stripe replied:
Alright, that appears to have worked - thanks! 👍

PiotrSikora commented on test/common/ssl/test_data/crl/crl_number:
```diff
@@ -0,0 +1 @@
+01
```
Do we need this file?

PiotrSikora commented on test/common/ssl/test_data/crl/crl_number.old:
```diff
@@ -0,0 +1 @@
+00
```
Do we need this file?

PiotrSikora commented on test/common/ssl/test_data/crl/index.txt.attr:
```diff
@@ -0,0 +1 @@
+unique_subject = no
```
Do we need this file?

PiotrSikora commented on test/common/ssl/test_data/crl/index.txt:
```diff
@@ -0,0 +1 @@
+R	191220171237Z	171220171254Z	D9CBF429EE201F1F	unknown	/C=US/ST=California/L=San Francisco/O=Lyft/OU=Lyft Engineering/CN=Revoked Certificate
```
Do we need this file?

PiotrSikora commented on test/common/ssl/test_data/not_a_crl.crl:
```diff
@@ -0,0 +1,3 @@
+-----BEGIN X509 CRL-----
```
Could you replace this with inline test?

adunham-stripe replied:
I don't think I can do that easily since the v1 config doesn't support inline data, AFAIK.  Happy to write out a temporary file during the test, perhaps?

PiotrSikora replied:
It's fine to replace the existing v1 test with v2 test, since that's the recommended syntax anyway... 

Also, there should be v2 tests for success / invalid content (i.e. replacement for this file) / invalid file path.

adunham-stripe replied:
Added success / failure cases to the v2 tests - do you want me to just remove the v1 tests in `test/common/ssl/context_impl_test.cc`?

PiotrSikora replied:
Up to you, I'm fine either way.

PiotrSikora commented on test/common/ssl/test_data/certs.sh:
```diff
@@ -68,10 +69,24 @@ openssl req -new -x509 -days 730 -key selfsigned_key.pem -out selfsigned_cert.pe
 openssl req -new -key expired_key.pem -out expired_cert.csr -config selfsigned_cert.cfg -batch -sha256
 openssl x509 -req -days -365 -in expired_cert.csr -signkey expired_key.pem -out expired_cert.pem
 
+# Generate revoked_cert.pem
```
We could simply revoke one of the existing certs and add it to the CRL.

If you prefer `revoked_cert.pem`, that's fine with me, but perhaps reuse `san_with_dns.cfg` to produce cert instead of adding another copy of the same `.cfg` file?

adunham-stripe replied:
Good point - I just pushed a commit that revokes an existing cert and tweaks the tests.

PiotrSikora commented on test/common/ssl/test_data/certs.sh:
```diff
@@ -68,10 +69,24 @@ openssl req -new -x509 -days 730 -key selfsigned_key.pem -out selfsigned_cert.pe
+
+# Revoke the certificate and generate a CRL
+openssl ca -revoke revoked_cert.pem -keyfile ca_key.pem -cert ca_cert.pem -config crl.cfg
+openssl ca -gencrl -keyfile ca_key.pem -cert ca_cert.pem -out revoked_cert.crl -config crl.cfg
```
`s/revoked_cert.crl/ca_cert.crl`

adunham-stripe commented:
@PiotrSikora - Okay, I rebased to fix a merge conflict, and added a couple commits that address almost all of your feedback.  I'm going to test the `X509_V_FLAG_CRL_CHECK_ALL` flag now, though the rest of the changes should be good for re-review.

adunham-stripe commented:
Well, after much testing, it appears that I can't reproduce the previous error I was seeing, so I just pushed a commit to add the `X509_V_FLAG_CRL_CHECK_ALL` flag.

PiotrSikora commented on source/common/ssl/context_config_impl.cc:
```diff
@@ -55,6 +55,8 @@ ContextConfigImpl::ContextConfigImpl(const envoy::api::v2::CommonTlsContext& con
       private_key_path_(config.tls_certificates().empty()
                             ? ""
                             : getDataSourcePath(config.tls_certificates()[0].private_key())),
+      certificate_revocation_list_(readDataSource(config.validation_context().crl(), true)),
```
Erm, I meant to move the `certificateRevocationList` / `certificateRevocationListPath` handling in all the files, not just this one (i.e. you should also move all the new code in `context_config.h`, `tls_context_json.cc`, `context_config_impl.h` and `context_impl.cc`).

adunham-stripe replied:
Ah, gotcha - moved the rest!

PiotrSikora commented on test/common/ssl/test_data/revoked_cert.cfg:
```diff
@@ -0,0 +1,36 @@
+[req]
```
This file is no longer being used.

PiotrSikora commented on source/common/ssl/context_impl.cc:
```diff
@@ -85,6 +85,30 @@ ContextImpl::ContextImpl(ContextManagerImpl& parent, Stats::Scope& scope,
+          fmt::format("Failed to load CRL from {}", config.certificateRevocationListPath()));
+    }
+
+    X509_STORE* store_ctx = SSL_CTX_get_cert_store(ctx_.get());
```
Sorry, I know that I suggested it, but this should be actually `store`, not `store_ctx`.

PiotrSikora replied:
Also, I know that's not your code (and I cannot comment on it), but in the CA handling, there is a call to `SSL_CTX_get_cert_store()` on each iteration, instead of before the loop, like you're doing here.

Could you update the old code to match yours? Thanks!

PiotrSikora commented on test/common/ssl/test_data/certs.sh:
```diff
@@ -68,10 +68,19 @@ openssl req -new -x509 -days 730 -key selfsigned_key.pem -out selfsigned_cert.pe
 
 rm *csr
 rm *srl
+rm -r crl_*
```
You don't need `-r` anymore.

PiotrSikora commented on source/common/ssl/context_config_impl.cc:
```diff
@@ -43,6 +43,8 @@ ContextConfigImpl::ContextConfigImpl(const envoy::api::v2::CommonTlsContext& con
       ca_cert_(readDataSource(config.validation_context().trusted_ca(), true)),
       ca_cert_path_(getDataSourcePath(config.validation_context().trusted_ca())),
+      certificate_revocation_list_(readDataSource(config.validation_context().crl(), true)),
+      certificate_revocation_list_path_(getDataSourcePath(config.validation_context().crl())),
```
There is one more thing missing here. You should throw an exception (down in the function body) if CRL is set, but there is no CA. All v1 tests are doing this, which obviously shouldn't work.

adunham-stripe replied:
Yeah, good point - I've pushed a commit that adds this, along with some tests.

PiotrSikora commented on source/common/ssl/context_config_impl.cc:
```diff
@@ -66,6 +68,9 @@ ContextConfigImpl::ContextConfigImpl(const envoy::api::v2::CommonTlsContext& con
   // TODO(htuch): Support multiple hashes.
   ASSERT(config.validation_context().verify_certificate_hash().size() <= 1);
+  if (ca_cert_.empty() && !certificate_revocation_list_.empty()) {
+    throw EnvoyException("Cannot have a CRL without a CA certificate");
```
Sigh, it looks that this comment got lost somehow.

This should be:

        throw EnvoyException(fmt::format("Failed to load CRL from {} without trusted CA certificates",
                                         config.certificateRevocationListPath()));

to match other errors.

adunham-stripe replied:
@PiotrSikora - Hah, no worries!  Just pushed a commit that changes this and updates the tests.

PiotrSikora commented on test/common/ssl/context_impl_test.cc:
```diff
@@ -284,5 +285,45 @@ TEST_F(SslServerContextImplTicketTest, TicketKeyInlineStringFailTooSmall) {
+  )EOF";
+
+  EXPECT_THROW_WITH_REGEX(loadConfigJson(json), EnvoyException,
+                          "Failed to load CRL from .*/not_a_crl.crl$");
```
Nit: missing `^` to indicate start of the message (you're using it in other places).

adunham-stripe commented:
@PiotrSikora - Nit fixed, and comment added!

PiotrSikora commented on test/common/ssl/test_data/README.md:
```diff
@@ -2,7 +2,8 @@
 - **CA**: Certificate Authority for **No SAN**, **SAN With URI** and **SAN With
   DNS**. It has the self-signed certificate *ca_cert.pem*. *ca_key.pem* is its
-  private key.
+  private key.  Additionally, we create a CRL for this CA (*ca_cert.crl*) that
```
One space should be enough for everybody ;)

PiotrSikora replied:
FWIW, this also breaks **ci/circleci: format** check.

adunham-stripe replied:
Aaaaaaagh 😛 

Okay, fixed.

mattklein123 commented:
@adunham-stripe sorry one last thing, it would be cool if you could do a quick doc PR to this page: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/ssl to mention that we now support CRLs (with link to config option). Thank you if you have time. cc @PiotrSikora

adunham-stripe commented:
Thanks very much for helping get this merged, everyone!  I submitted the docs PR over at https://github.com/envoyproxy/data-plane-api/pull/437
